keps/sig-cloud-provider/20191004-out-of-tree-credential-providers.md

```diff
@@ -0,0 +1,220 @@
+last-updated: 2019-12-10
+status: provisional
+---
+
+# Out-of-Tree Credential Providers
+
+## Table of Contents
+
+<!-- toc -->
+- [Release Signoff Checklist](#release-signoff-checklist)
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [External Credential Provider](#external-credential-provider)
+  - [Alternatives Considered](#alternatives-considered)
+    - [API Server Proxy](#api-server-proxy)
+    - [Sidecar Credential Daemon](#sidecar-credential-daemon)
+    - [Bound Service Account Token Flow](#bound-service-account-token-flow)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Implementation History](#implementation-history)
+- [Infrastructure Needed](#infrastructure-needed)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+Replace the existing in-tree container image registry (registry) credential providers with an external and pluggable credential provider mechanism and remove in-tree credential providers.
+
+## Motivation
+
+kubelet uses cloud provider specific SDKs to obtain credentials when pulling container images from cloud provider specific registries. The use of cloud provider specific SDKs from within the main Kubernetes tree is deprecated by [KEP-0002](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cloud-provider/20180530-cloud-controller-manager.md) and all existing uses need to be migrated out-of-tree. This KEP supports that migration process by removing this SDK usage.
+
+In addition to supporting cloud provider migration this KEP supports allowing users to support multiple container registries across different cloud providers by introducing a pluggable interface to cloud specific credential providers such that a single user could run multiple credential providers for different clouds at the same time. Currently this functionality is gated by having only a single, active cloud provider within the API server and kubelet.
+
+### Goals
```

Comment:
## Is it a goal to continue to support programmatic access?

The current `credentialprovider` package is depended on by [ggcr](https://github.com/google/go-containerregistry) (github.com/google/go-containerregistry) and, indirectly, by the the consumers of ggcr. This package has become difficult to consume, hence https://github.com/kubernetes/kubernetes/issues/82396.

If programmatic access is dropped, ggcr and the projects based on it will need to include cloud provider specific binaries in their releases, which will complicate the packaging of these releases and make them much less convenient to install.

It may be tempting to limit this KEP to kubelet's needs, but some of the above projects provide important infrastructure in support of Kubernetes. For example, an [image relocation project](https://github.com/pivotal/image-relocation) based on ggcr is used to copy images to a private registry - useful when running Kubernetes in disconnected environments.

Reply:
FWIW both Knative and Tekton build on this.

Reply:
Isn't this exactly why we discourage depending on kubernetes/kubernetes? Since we want to be able to make design decisions based on what is best for Kubernetes, not the library. This seems like it would be a non-goal to me.

That said, we could take steps to make this less painful. By providing a common interface or reference implementation, it would be a lot easier to pull the credential providers together into a single library (assuming they're opensource and implemented in go, using the provided interface).

Reply:
I think I agree with @tallclair.  I added "Continuing to support projects that import the credential provider package."  as a non-goal.  Once the cloud provider specific code is outside kubernetes/kubernetes, they can choose to import each cloud provider package or using the exec-based interface kubelet will use.  We can publish an interface and encourage providers to use it to make programmatic usage easier, but I don't know that needs to be included in the KEP.

Reply:
> We can publish an interface and encourage providers to use it to make programmatic usage easier

For what it's worth, this would satisfy my needs.